crates/uv/tests/lock.rs

```diff
@@ -965,3 +965,320 @@ fn lock_git_sha() -> Result<()> {
 
     Ok(())
 }
+
+/// Lock a requirement from PyPI, respecting the `Requires-Python` metadata.
+#[test]
+fn lock_requires_python() -> Result<()> {
+    let context = TestContext::new("3.12");
+
+    // Require >=3.7, which is incompatible with newer versions of `pygls`.
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r#"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        requires-python = ">=3.7"
+        dependencies = ["pygls>=1.1.0"]
+        "#,
+    )?;
+
+    uv_snapshot!(context.filters(), context.lock(), @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+
+    ----- stderr -----
+    warning: `uv lock` is experimental and may change without warning.
+      × No solution found when resolving dependencies:
+      ╰─▶ Because the requested Python version (>=3.7) does not satisfy Python>=3.8 and the requested Python version (>=3.7) does not satisfy Python>=3.7.9,<3.8, we can conclude that Python>=3.7.9 are incompatible.
```

Comment:
Not a particularly clear message but it's roughly the same as `echo "pygls>=1.1.0" | cargo run pip compile --python-version 3.7 -` before any of my changes.

Reply:
Here's Poetry:

```
The current project's Python requirement (>=3.7) is not compatible with some of the required packages Python requirement:
  - pygls requires Python >=3.7.9,<4, so it will not be satisfied for Python >=3.7,<3.7.9 || >=4
  - pygls requires Python >=3.7.9,<4, so it will not be satisfied for Python >=3.7,<3.7.9 || >=4
  - pygls requires Python >=3.7.9,<4, so it will not be satisfied for Python >=3.7,<3.7.9 || >=4
  - pygls requires Python >=3.7.9,<4, so it will not be satisfied for Python >=3.7,<3.7.9 || >=4
  - pygls requires Python >=3.7.9,<4, so it will not be satisfied for Python >=3.7,<3.7.9 || >=4
  - pygls requires Python >=3.8, so it will not be satisfied for Python >=3.7,<3.8
  - pygls requires Python >=3.8, so it will not be satisfied for Python >=3.7,<3.8

Because no versions of pygls match >1.1.0,<1.1.1 || >1.1.1,<1.1.2 || >1.1.2,<1.2.0 || >1.2.0,<1.2.1 || >1.2.1,<1.3.0 || >1.3.0,<1.3.1 || >1.3.1
 and pygls (1.1.0) requires Python >=3.7.9,<4, pygls is forbidden.
And because pygls (1.1.1) requires Python >=3.7.9,<4, pygls is forbidden.
And because pygls (1.1.2) requires Python >=3.7.9,<4
 and pygls (1.2.0) requires Python >=3.7.9,<4, pygls is forbidden.
And because pygls (1.2.1) requires Python >=3.7.9,<4
 and pygls (1.3.0) requires Python >=3.8, pygls is forbidden.
So, because pygls (1.3.1) requires Python >=3.8
 and poetry-test depends on pygls (>=1.1.0), version solving failed.

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties

    For pygls, a possible solution would be to set the `python` property to ">=3.7.9,<4"
    For pygls, a possible solution would be to set the `python` property to ">=3.7.9,<4"
    For pygls, a possible solution would be to set the `python` property to ">=3.7.9,<4"
    For pygls, a possible solution would be to set the `python` property to ">=3.7.9,<4"
    For pygls, a possible solution would be to set the `python` property to ">=3.7.9,<4"
    For pygls, a possible solution would be to set the `python` property to ">=3.8"
    For pygls, a possible solution would be to set the `python` property to ">=3.8"
```

Reply:
We should add a hint here that "requested Python version" refers to the version you declared in your pyproject.toml and that - unlike everything else - that version needs to stricter than the one in your dependencies.

Reply:
Will do in a separate PR.